api/http/api_inventory.go

```diff
@@ -772,50 +772,60 @@ func getTenantContext(ctx context.Context, tenantId string) context.Context {
 }
 
 func (i *inventoryHandlers) InternalDevicesStatusHandler(w rest.ResponseWriter, r *rest.Request) {
-	ctx := r.Context()
+	const (
+		StatusDecommissioned = "decommissioned"
+		StatusAccepted       = "accepted"
+		StatusRejected       = "rejected"
+		StatusPreauthorized  = "preauthorized"
+		StatusPending        = "pending"
+	)
+	var (
+		ids    []model.DeviceID
+		result *model.UpdateResult
+	)
 
+	ctx := r.Context()
 	l := log.FromContext(ctx)
 
-	status := r.PathParam("status")
-	if status == "" {
-		u.RestErrWithLog(w, r, l, errors.New("status not provided"), http.StatusBadRequest)
-		return
-	}
+	tenantID := r.PathParam("tenant_id")
+	ctx = getTenantContext(ctx, tenantID)
 
-	tenantId := r.PathParam("tenant_id")
-
-	ctx = getTenantContext(ctx, tenantId)
+	status := r.PathParam("status")
 
-	var ids []string
 	err := r.DecodeJsonPayload(&ids)
 	if err != nil {
 		u.RestErrWithLog(w, r, l, errors.Wrap(err, "cant parse device ids"), http.StatusBadRequest)
 		return
 	}
 
-	for _, id := range ids {
-		attrs := model.DeviceAttributes{
-			{
-				Name:        "status",
-				Description: nil,
-				Value:       status,
-				Scope:       model.AttrScopeIdentity,
-			},
-		}
+	switch status {
+	case StatusAccepted, StatusPreauthorized, StatusPending:
+		// Update statuses
+		attrs := model.DeviceAttributes{{
+			Name:  "status",
+			Scope: model.AttrScopeIdentity,
+			Value: status,
+		}}
+		result, err = i.inventory.UpsertDevicesAttributes(ctx, ids, attrs)
 
-		if status == "decommissioned" {
-			err = i.inventory.DeleteDevice(ctx, model.DeviceID(id))
-		} else {
-			//upsert the attributes
-			err = i.inventory.UpsertAttributes(ctx, model.DeviceID(id), attrs)
-		}
-		if err != nil {
-			u.RestErrWithLogInternal(w, r, l, err)
-			return
-		}
+	case StatusDecommissioned, StatusRejected:
```

Comment:
where did the `StatusRejected` come from?

`decommissioned` is a very special case (rejecting preauth'd authsets) and it's one case when we want to get rid of the device.

we don't want the get rid of all device data if it's rejected.

Reply:
FYI there's a related discussion going on under a bug I'm working on:
https://tracker.mender.io/browse/MEN-2558?focusedCommentId=108637&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-108637

note that we stick to the philosophy of not removing devices even if there are no authsets; 
we'll possibly go as far as introducing a new `noauth` state just to keep a device around.

that decision was made a long long time ago, but Eystein just reiterated it.

Reply:
Thanks for sharing, I was not aware of this and understood that we should only keep the rejected entry in deviceauth (after all, the device has no way of updating the inventory once it's rejected). I'll move the rejected status to the previous case then.

Reply:
thank you! yes, it's confusing to me as well, I mean i couldn't recite all those rules concerning the authset/device lifecycle off the cuff...
anyway, good that we're having the discussion, I wouldn't have spotted it otherwise (or deemed it too important).